### PR TITLE
Extend OpenVSwitch modules

### LINKED ISSUE #58986: [FEATURE REQUEST]  Extend OpenVSwitch modules with VLAN and DB functions
The OpenVSwitch modules do currently not provide functions for adding bridges that are children of an existing bridge and use a specific VLAN tag. They do not provide functions for setting arbitrary properties (e.g. setting a fixed MAC address for an interface) either.

Therefore the following changes are needed:

- Modify `salt.states.openvswitch_bridge.present(…)`, adding two new optional parameters for the parent bridge and VLAN ID.
- Modify `salt.modules.openvswitch.bridge_create(…)`, so that it takes two new optional parameters for the parent bridge and VLAN ID. This function is used by `salt.states.openvswitch_bridge.present(…)`.
- Add a function `bridge_to_parent(…)` to `salt.modules.openvswitch`. This function is used to determine the parent of an existing bridge.
- Add a function `bridge_to_vlan(…)` to `salt.modules.openvswitch`. This function is used to determine the VLAN ID of an existing bridge.
- Add a module `salt.states.openvswitch_db` with a `managed(…)` function. This function can be used to make sure that a record in the OpenVSwitch database has certain values.
- Add a function `db_get` to `salt.modules.openvswitch`. This function is used to read a value from the OpenVSwitch database.
- Add a function `db_set` to `salt.modules.openvswitch`. This function is used to write a value to the OpenVSwitch database.

All these changes have been accepted into Salt before with PR #50564. However, this PR was made against the develop branch and they never got merged into master and thus never made it into a release.

I have now ported these changes to the current master and also made adjustments to make them fit the current guidelines (for example I added tests for all added or modified functions).

I will create a new PR with these changes soon. This issue has only been created to have an issue ID that can be referenced in the changelog entry.

### PR DESCRIPTION
**Pull request by smarsching**
### What does this PR do?

This PR adds functions to the OpenVSwitch modules for two purposes:

- Creating bridges that have a parent bridge. Such bridges can be used in OpenVSwitch for having a bridge that is associated with a specific (tagged) VLAN on a parent bridge.
- Updating values in the OpenVSwitch database. This can be used to update arbitrary properties of OpenVSwitch objects. For example, it can be used to assign a fixed MAC address to the interface associated with a certain OpenVSwitch port.

In detail, the following changes are made to the code:

- The `openvswitch_bridge.present(…)` state function is modified, adding two new optional parameters for the parent bridge and VLAN ID.
- The `openvswitch.bridge_create(…)` execution function is modified, adding two new optional parameters for the parent bridge and VLAN ID. This function is used by `salt.states.openvswitch_bridge.present(…)`.
- The `openvswitch.bridge_to_parent(…)` execution function is added. This function is used by `salt.states.openvswitch_bridge.present(…)` to determine the parent of an existing bridge.
- The `openvswitch.bridge_to_vlan(…)` execution function is added. This function is used by `salt.states.openvswitch_bridge.present(…)` to determine the VLAN ID of an existing bridge.
- The `openvswitch_db` state modules is added. This module has a single function called `openvswitch_db.managed`. This function can be used to make sure that a record in the OpenVSwitch database has certain values.
- The `openvswitch.db_get` execution function is added. This function is used by `openvswitch_db.managed(…)` to read a value from the OpenVSwitch database.
- The `openvswitch.db_set` execution function is added. This function is used by `openvswitch_db.managed(…) to write a value to the OpenVSwitch database.

### What issues does this PR fix or reference?
Fixes: #58986
There was an earlier PR (#50564) essentially providing the same changes that got merged into `develop`, but was never ported to `master`.

### Merge requirements satisfied?
- [X] Docs
- [X] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
No

### CODE REVIEW

**Review by dmurphy18 — CHANGES_REQUESTED** [review]
@smarsching Are you still interested in this PR ?, if unresponsive the PR shall be closed after several days.
Test cases need addressing as well as resolving conflicts

**Comment by smarsching** [comment]
@dmurphy18 I rebased the PR against the current `master` branch and addressed all linting issues. Are there any other things that need to be changed?

I saw that two checks (pr-centos-7-x86_64-py3-pycryptodome-pytest and pr-centos-7-x86_64-py3-tcp-pytest) failed, but the log output does not give any information about why they failed (at least no information that I could decipher). Could you have a look please and tell me, so that I can try to reproduce and fix these problems if needed?

**Comment by dmurphy18** [comment]
@smarsching Thanks for the updates,  will take a look and let you know what I find

**Comment by smarsching** [comment]
@dmurphy18 Thanks for having another look. It seems like merging the `master` branch triggered another run of the checks, which all succeeded now. Maybe the two failed checks were simply causes by a temporary hiccup in the CI system.

Please let me know whether any more actions are needed before this can be merged.

File: salt/states/openvswitch_bridge.py [hunk 1]
```diff
@@ -12,12 +12,16 @@ def __virtual__():
...
 
 
-def present(name):
+def present(name, parent=None, vlan=None):
     """
     Ensures that the named bridge exists, eventually creates it.
 
     Args:
         name: The name of the bridge.
+        parent: The name of the parent bridge (if the bridge shall be created
```
Comment (dmurphy18) [inline salt/states/openvswitch_bridge.py:21]: These new additional parameters could use a versionadded see https://docs.saltproject.io/en/latest/topics/development/conventions/style.html
```
.. versionadded:: 3006
```
see above about dropping the 'A', that is 
```
   name : string
     bridge name
```

File: salt/states/openvswitch_db.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    Only make these states available if Open vSwitch module is available.
+    """
+    return "openvswitch.db_get" in __salt__
+
+
+def managed(name, table, data, record=None):
+    """
+    Ensures that the specified columns of the named record have the specified
+    values.
+
```
Comment (dmurphy18) [inline salt/states/openvswitch_db.py:17]: see style comments above

**Review by dmurphy18 — CHANGES_REQUESTED** [review]
Have some style changes
The test files are the old test/unit and while still valid, wondering if you would be able to make them pytests since that is what is preferred moving forward ?

**Comment by smarsching** [comment]
Thank you very much for your review and your comments regarding the style. I will address the changes that you requested, but it will probably take a few days before I find the time to do so. I will notify you when the code is ready for another review.

**Comment by Ch3LL** [comment]
We are attempting to get everything merged in by this Friday. Just letting you know if you want to get this in before then.

**Comment by smarsching** [comment]
@Ch3LL Thanks for letting me know. Unfortunately, I couldn’t get to it earlier this week due to other, more urgent projects. Maybe you still have a change to merge it today.

@dmurphy18 I addressed your change requests regarding documentation style. All functions that have been newly introduced or that have had parameters added now use the newer style. I also added “versionadded” to all newly introduced modules, functions, and parameters. The unit tests have been converted to the new “pytest” style. I am now waiting for all the checks to complete and hope that I have not introduced any new problems with these changes. 😬

**Comment by smarsching** [comment]
It looks like all the “-pytest” checks failed, but with a quite unspecific message “script returned exit code 20”. Unfortunately, the build log is truncated and thus does not contain any useful information either, so now I am not sure whether these failures are caused by my changes (converting the tests to pytest) or whether they appeared because I rebased the branch on the newest version of the master branch.

@dmurphy18 Do you have any idea how I could get a more concrete error message from this jobs that might at least give me a hint to the cause of the problems?

**Comment by smarsching** [comment]
After managing to create a local test setup, I got a meaningful error message that allowed me to figure out what’s happening: The tests that were converted to pytest should also have been moved to `tests/pytest`. After doing this, the tests run, but there still is a problem with them. As soon as I have fixed this (probably in an hour or so), I will push the changes. With a bit of luck, this will resolve all issues and make the checks succeed.

**Comment by dmurphy18** [comment]
FYI - @waynew Holds Test Clinics on Twitch, which issues writing tests for Salt https://www.twitch.tv/saltprojectoss

Salt now only accepts new tests if they are pytests. Can still modify / update existing unit tests, but these are being transitioned to pytest, as time allows (background task).

Thanks for moving your tests over to pytest, which as you found reside under tests/pytest.  There are also helpers for pytest with Salt, see https://pytest-salt-factories.readthedocs.io/en/stable/contents.html and https://pytest-skip-markers.readthedocs.io/en/latest/contents.html#table-of-contents

**Comment by smarsching** [comment]
@dmurphy18 Thanks for your reply. I think that I managed to sort out the problems with the tests. All checks except for the “pr-windows-2019-x64-py3-pytest” check succeeded now. That check is still in progress. It looks to me like it cannot progress due to some problem with an EC2 instance.

As it seems like all issues have been resolved, could you please give this PR another review, so that it can be merged?

**Comment by smarsching** [comment]
@dmurphy18 Just so that you know, the last check has completed as well and all checks are marked as successful now.

**Comment by dmurphy18** [comment]
Thanks @smarsching , will review soon

**Review by dmurphy18 — APPROVED** [review]

**Review by garethgreenaway — APPROVED** [review]

**Pull request merged**